### ISSUE TITLE
Mac M1 (arm) Python support

### ISSUE THREAD

**Issue by Alex-Vasile** [alphacep/vosk-api#692]
I tried installing using pip but had no luck, and there doesn't seem to be a built wheel under the releases.

Is there support for the arm

Am I going to have to figure out how to build it all from source?

**Comment by nshmyrev** [comment]
> Is there support for the arm

No

> Am I going to have to figure out how to build it all from source?

Yes. I don't think we have access to any M1 build environment right now.

Related is #681

**Comment by cosbor11** [comment]
+1
For me when trying to install on Mac with M1, 

when running: 
```
pipenv global 3.8.12
pip install vosk
```

I get the following error: 
```
ERROR: Could not find a version that satisfies the requirement vosk (from versions: none)
ERROR: No matching distribution found for vosk
```

**Comment by nshmyrev** [comment]
A common option (scipy uses it for example) is to run python x86 in Rosetta2 terminal instead. For example:

https://github.com/openvinotoolkit/openvino_notebooks/discussions/10

**Comment by rolkon** [comment]
I got VOSK working on my MacBook M1, with Rosetta. I mainly work within jupyter notebooks.

My steps:

1. I installed a second, x86 version of homebrew on my machine (according to [https://stackoverflow.com/a/64951025](https://stackoverflow.com/a/64951025)). I also named mine `brow`, like in the tutorial.
2. I then used `brow install python3.x` to install Python
3. I used pip for all subsequent libraries I needed. Every command which addresses the x86 version of Python now needs to be prefixed with `ib`, or any other prefix that you chose in step 1. So for vosk, this means `ib pip3 install vosk`
4. I also installed jupyter notebook with `ib pip3 install notebook`
5. I started the notebook with `ib jupyter-notebook` and was able to import and work with the vosk library without a problem.

This solution means that now you have two completely separated instances of brew, as well as two stacks of python and libraries. However, they are nicely isolated and self-contained, and you can easily remove them without effecting the other. It also makes it very easy to later migrate code from Rosetta x86 to the M1 - you just have to install the libraries with the right pip, and launch the right notebook.

**Comment by ricardowiseverge** [comment]
How did you manage running tests in an IDEA for example running in PyCharm? It gets confused and doesn't find the packages.

Also on the kernel part even if I start `ib jupyter lab` the kernel doesn't feels to be right. Tried to understand that using which python and conda info but couldn't find what could be wrong

Besides that I'm having issues in running other things. When I do `conda info` I get this:

![image](https://user-images.githubusercontent.com/91593162/154668295-99a49f83-6d07-427b-a31f-ea9487ed832e.png)

**Comment by aivoicesystems** [comment]
I'd like to get libvosk compiled natively on Apple silicon (started this on #905).

With a little bit of tweaking I can compile kaldi and vosk-api src natively on my M1 but there are a few things I'd like to sort out.

1. I think the goal should be that the Darwin version of libvosk (libvosk.dylib) should be a universal binary - including both x86 and arm binaries in the one library. Is there interest in going this far?
2. If that is done the aim should be for the "standard" versions of the library (for example the jar on maven central in the case of the java version) to include this universal binary. I don't know anything about the distribution process.
3. There are a few minor problems with the build setup as it stands now that could be cleaned up to smooth compilation on a mac. Is there interest in doing that?

Thanks!

**Comment by dynodino** [comment]
@aivoicesystems Can you share the commands you used to compile and install?

**Comment by aivoicesystems** [comment]
@dynodino I have put a script that implements the logic of Docker.manylinux on #905.

Here is an updated version of the script that builds both kaldi and vosk-api.

To run it you need a case-sensitive file system (beware this is NOT the default configuration for a mac os fs) and a fairly standard set of tools (I use homebrew). The script checks both these before trying the build.

Let me know if you have any other questions.

[build-macos.sh.zip](https://github.com/alphacep/vosk-api/files/8505410/build-macos.sh.zip)

**Comment by aivoicesystems** [comment]
> To run it you need a case-sensitive file system (beware this is NOT the default configuration for a mac os fs)

It turns out that working around this is fairly straight forward. I've reworked the script so it does NOT require a case-sensitive file system anymore.

[vosk-macos.sh.zip](https://github.com/alphacep/vosk-api/files/8509882/vosk-macos.sh.zip)

**Comment by sskorol** [comment]
@aivoicesystems testing the script... Wondering why did you omit the following kaldi features: `online2 lm rnnlm`? They are all present in the original scripts. I also try to adopt it to Docker, but have no luck yet. Will check the latest update.

**Comment by aivoicesystems** [comment]
> Wondering why did you omit the following kaldi features: online2 lm rnnlm?

Simple reason - I missed them. I'm testing the build now with this change:

```
114c114
<   make -j ${nproc}
---
>   make -j ${nproc} online2 lm rnnlm
```

**Comment by aivoicesystems** [comment]
That builds fine

**Comment by sskorol** [comment]
@aivoicesystems tried to build it using the latest script. Several notes:
- initially, it failed to find the following dependencies: `aclocal autoreconf cmake`
- then I had some other failure which required `automake` to be installed
- then I had another failure locating `glibtoolize`
- so I ended up installing the following dependencies: `brew install autoconf cmake automake libtool`

Also, I see this script doesn't install python module. So I added the following after making vosk-api:
```shell
cd ../python
python3 setup.py install
```
The build is completed. Started testing it and faced the following issues:
- python samples don't work out of the box
- initial failure was related to the missing cifi module; had to explicitly install it
- then I wasn't able to run the py code because it was unable to find [libvosk.dyld](https://github.com/alphacep/vosk-api/blob/master/python/vosk/__init__.py#L17); obviously, because the script's output is `libvosk.dylib` but [setup.py](https://github.com/alphacep/vosk-api/blob/master/python/setup.py#L55) bundles `.dyld` ext
- then I tested if it'd work if I rename the expected ext to `dylib` directly in py sources and put the newly generated file into vosk package; it worked
- lastly, I just changed the following line in a build script: `EXT=dyld KALDI_ROOT=../../kaldi make -j ${nproc}` (apart from cp and final logs output) and now it works as expected.

So to summarize:
- I'd be more specific about which tools are required to be installed, as `aclocal autoreconf` are parts of the other packages, and not really sufficient.
- python bundle part is missing
- python module requires a different libvosk ext

Other than that, it's also interesting to get some stats on total build time.

**Comment by sskorol** [comment]
FYI, just published M1 (arm64) docker images:
- [api](https://hub.docker.com/r/sskorol/vosk-api/tags)
- [server](https://hub.docker.com/r/sskorol/vosk-server/tags)

**Comment by aivoicesystems** [comment]
@sskorol That's progress.

Do you have the changes to the script that you mentioned?

Regarding the python build, I'm not using the python version (I'm only in this issue because a more general issue about Apple Silicon (#905) I opened got closed and referred to here) but happy that the build can be used for python. What about builds for other languages? How are they meant to be managed? For example, how do we get an M1 version into the jar on maven central for java users?

Just out of interest is there a reason to use .dyld rather than .dylib as the file extension? The mac standard seems to be .dylib.

**Comment by sskorol** [comment]
> Do you have the changes to the script that you mentioned?

@aivoicesystems [here](https://github.com/sskorol/vosk-api-gpu/blob/main/build-m1-local.sh) I've just added a required python part. Left credits for you in a README.

The reason why I add Python is the official set of docker images that are [python-oriented](https://github.com/alphacep/vosk-server/blob/master/docker/Dockerfile.kaldi-vosk-server#L52-L56).

> What about builds for other languages?

The other languages just wrap the native code the same way it's done for Python. That's e.g. [Java](https://github.com/alphacep/vosk-api/blob/master/java/lib/src/main/java/org/vosk/LibVosk.java). Publishing is handled by a [gradle plugin](https://github.com/alphacep/vosk-api/blob/master/java/lib/build.gradle#L32-L61).

I'm not a repo owner. Such questions are better to ask @nshmyrev. As well as why `.dyld` is used instead of `.dylib`. I'm the same curious user as you are.

**Comment by aivoicesystems** [comment]
Thanks @sskorol, that all makes sense.

I'll let @nshmyrev have a say about the builds for other languages and the .dyld naming.

**Comment by iworkinpixels** [comment]
> @dynodino I have put a script that implements the logic of Docker.manylinux on #905.
> 
> Here is an updated version of the script that builds both kaldi and vosk-api.
> 
> To run it you need a case-sensitive file system (beware this is NOT the default configuration for a mac os fs) and a fairly standard set of tools (I use homebrew). The script checks both these before trying the build.
> 
> Let me know if you have any other questions.
> 
> [build-macos.sh.zip](https://github.com/alphacep/vosk-api/files/8505410/build-macos.sh.zip)

Your script worked wonderfully, as far as building the dylib, so thank you for that!
But... forgive me, for I am very VERY stupid... what do I need to do once I have the compiled .dylib? Where do I need to link it to so that "import vosk" will work inside my python example script?

**Comment by sskorol** [comment]
@iworkinpixels [here's](https://github.com/sskorol/vosk-api-gpu/blob/main/m1/build-local.sh) the same version, but with Python wrapper. Just make sure you activated your venv before running. Otherwise, it'll install the package globally. There's also a [docker build](https://github.com/sskorol/vosk-api-gpu/blob/main/m1/build.sh) available.

**Comment by nshmyrev** [comment]
Hi. We have just published official universal package, it should support M1:

https://pypi.org/project/vosk/0.3.38/#files

also here:

https://github.com/alphacep/vosk-api/releases/tag/v0.3.38

**Comment by PaulMest** [comment]
> Hi. We have just published official universal package, it should support M1:
> 
> https://pypi.org/project/vosk/0.3.38/#files
> 
> also here:
> 
> https://github.com/alphacep/vosk-api/releases/tag/v0.3.38

This doesn't seem to work for me...

```sh
$ pip install vosk-0.3.38-py3-none-macosx_10_6_universal.whl
ERROR: vosk-0.3.38-py3-none-macosx_10_6_universal.whl is not a supported wheel on this platform.
```

System details:
* macOS Monterey 12.4 (21F79)
* MacBook Pro (16-inch, 2021
* Apple M1 Max
* Python 3.10.4

**Comment by sskorol** [comment]
That's because macOS version is hardcoded https://github.com/alphacep/vosk-api/blob/master/python/setup.py. It should be detected automatically and depend on a platform. I faced the same before while local wheel building. For M1 Monterey package suffix should be `macosx_12_0_arm64`.

**Comment by nshmyrev** [comment]
@PaulMest what is the full output of `pip3 debug --verbose` on your system.

**Comment by nshmyrev** [comment]
And also `pip3 install vosk-0.3.38-py3-none-macosx_10_6_universal.whl` with pip3 instead of pip.

**Comment by PaulMest** [comment]
@sskorol Renaming the file appears to have worked!

```sh
$ pip install vosk-0.3.38-py3-none-macosx_12_0_arm64.whl

Processing vosk-0.3.38-py3-none-macosx_12_0_arm64.whl
Collecting cffi>=1.0
  Downloading cffi-1.15.0-cp310-cp310-macosx_11_0_arm64.whl (173 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 174.0/174.0 kB 3.3 MB/s eta 0:00:00
Collecting pycparser
  Downloading pycparser-2.21-py2.py3-none-any.whl (118 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 118.7/118.7 kB 4.7 MB/s eta 0:00:00
Installing collected packages: pycparser, cffi, vosk
Successfully installed cffi-1.15.0 pycparser-2.21 vosk-0.3.38

$ (video-analysis-3.10.4) ➜  ~/code/video-analysis pip list
Package    Version
---------- -------
cffi       1.15.0
pip        22.1.1
pycparser  2.21
setuptools 62.3.2
vosk       0.3.38
```

**Comment by nshmyrev** [comment]
I uploaded universal2 variant, that should work.

**Comment by pzula** [comment]
I am using mac os big sir version 11.6.5  with an apple m1  chip and I'm unable to install ask using this command: `pip3 install vosk-0.3.38-py3-none-macosx_10_6_universal.whl`.  am I doing something wrong?

` pip3 install vosk-0.3.38-py3-none-macosx_10_6_universal.whl
WARNING: Requirement 'vosk-0.3.38-py3-none-macosx_10_6_universal.whl' looks like a filename, but the file does not exist
ERROR: vosk-0.3.38-py3-none-macosx_10_6_universal.whl is not a supported wheel on this platform.`

**Comment by nshmyrev** [comment]
The wheel filename is now universal2, not universal. You have to download it first if you want to install from wheel file. Or you can just run `pip3 install vosk`
